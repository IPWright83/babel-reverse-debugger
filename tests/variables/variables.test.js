const { exec } = require('child_process');
const path = require('path');

const runTransform = (inputFile) => {
    return new Promise((resolve, reject) => {
        const configFile = path.resolve(__dirname) + "/../../.babelrc-test";
        const command = `npx babel ${inputFile} --quiet --config-file ${configFile}`;
        exec(command, (error, stdout, stderr) => {
            if (error) {
                reject(error);
            }

            if (stderr) {
                reject(stderr.trim());
            }

            resolve(stdout.trim());
        })
    })
};

describe("Variable Decorators", () => {
    test.each([
        ["const assignment", "const-assignment.js"],
        ["let assignment", "let-assignment.js"],
        ["var assignment", "var-assignment.js"],
        ["multi value assignment", "multi-assignment.js"],
        ["String", "string-assignment.js"],
        ["Boolean", "bool-assignment.js"],
        ["Function", "function-assignment.js"],
        ["Arrow Function", "arrow-assignment.js"],
        ["Object", "object-assignment.js"],
        ["Re-assignment", "re-assignment.js"],
        ["Increment++", "incrementplusplus.js"],
        ["Increment+=", "incrementplusequals.js"],
        // ["Assign to Undefined", "assign-to-undefined.js"],
        // ["Assign to operation", "assign-to-operation.js"],
        // ["Assign to function result", "assign-to-func-result.js"],
        // ["Assign to built in function result", "assign-to-builtin-result.js"]
    ])("Converts %s correctly", async (name, inputFile) => {
        const filePath = path.resolve(__dirname, inputFile);
        const output = await runTransform(filePath);
        expect(output).toMatchSnapshot();
    });
});
