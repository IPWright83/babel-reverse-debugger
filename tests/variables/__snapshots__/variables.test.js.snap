// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Variable Decorators Converts Arrow Function correctly 1`] = `
""use strict";

var hello = ___captureAssignment("VariableDeclaration", "hello", 1, "\\u03BB", function () {
  ___instrumentFunction("ArrowFunctionExpression", "hello", 1, {});
  return ___instrumentReturn("ReturnStatement", 1, "Hello World");
});"
`;

exports[`Variable Decorators Converts Assign to Undefined correctly 1`] = `
""use strict";

var count = ___captureAssignment("VariableDeclaration", "count", 1, 3, 3);
count = ___captureAssignment("AssignmentExpression", "count", 2, undefined, undefined);;"
`;

exports[`Variable Decorators Converts Assign to built in function result correctly 1`] = `
""use strict";

var count = ___captureAssignment("VariableDeclaration", "count", 1, 3, 3);
count = ___captureAssignment("AssignmentExpression", "count", 2, "\\u03BB", Math.max(1, 2, 3, 4));;"
`;

exports[`Variable Decorators Converts Assign to function result correctly 1`] = `
""use strict";

var sum = ___captureAssignment("VariableDeclaration", "sum", 1, "\\u03BB", function (a, b) {
  ___instrumentFunction("ArrowFunctionExpression", "sum", 1, {
    a: a,
    b: b
  });
  return ___instrumentReturn("ReturnStatement", 1, a + b);
});
var result = ___captureAssignment("VariableDeclaration", "result", 2, "\\u03BB", sum(2, 4));"
`;

exports[`Variable Decorators Converts Assign to operation correctly 1`] = `
""use strict";

var count = ___captureAssignment("VariableDeclaration", "count", 1, 3, 3);
count = ___captureAssignment("AssignmentExpression", "count", 2, 3 + 7, 3 + 7);;"
`;

exports[`Variable Decorators Converts Boolean correctly 1`] = `
""use strict";

var isTrue = ___captureAssignment("VariableDeclaration", "isTrue", 1, true, true);"
`;

exports[`Variable Decorators Converts Chained assignment correctly 1`] = `
""use strict";

var value = ___captureAssignment("VariableDeclaration", "value", 1, "\\u03BB", 1 + 2 + 3 + 4);"
`;

exports[`Variable Decorators Converts Function correctly 1`] = `
""use strict";

var hello = ___captureAssignment("VariableDeclaration", "hello", 1, "\\u03BB", function () {
  ___instrumentFunction("FunctionExpression", "anonymous", 1, {});
  return ___instrumentReturn("ReturnStatement", 1, "Hello World");
});"
`;

exports[`Variable Decorators Converts Increment+= correctly 1`] = `
""use strict";

var count = ___captureAssignment("VariableDeclaration", "count", 1, 2, 2);
count += ___captureAssignment("AssignmentExpression", "count", 2, count + 3, count + 3);;"
`;

exports[`Variable Decorators Converts Mixed assignment correctly 1`] = `
""use strict";

var sum = ___captureAssignment("VariableDeclaration", "sum", 1, "\\u03BB", function (a, b) {
  ___instrumentFunction("ArrowFunctionExpression", "sum", 1, {
    a: a,
    b: b
  });
  return ___instrumentReturn("ReturnStatement", 1, a + b);
});
var value = ___captureAssignment("VariableDeclaration", "value", 2, "\\u03BB", 3 + sum(2, 1));"
`;

exports[`Variable Decorators Converts Object correctly 1`] = `
""use strict";

var foo = ___captureAssignment("VariableDeclaration", "foo", 1, {
  bar: {
    baz: "foo-bar-baz"
  }
}, {
  bar: {
    baz: "foo-bar-baz"
  }
});"
`;

exports[`Variable Decorators Converts Re-assignment correctly 1`] = `
""use strict";

var count = ___captureAssignment("VariableDeclaration", "count", 1, 0, 0);
count = ___captureAssignment("AssignmentExpression", "count", 2, 3, 3);;"
`;

exports[`Variable Decorators Converts Shorthand Operations correctly 1`] = `
""use strict";

var count = ___captureAssignment("VariableDeclaration", "count", 1, 1, 1);
count += ___captureAssignment("UpdateExpression", "count", 2, 1, count + 1), count;
count -= ___captureAssignment("UpdateExpression", "count", 3, 1, count - 1), count;
count *= ___captureAssignment("AssignmentExpression", "count", 4, count * 2, count * 2);;"
`;

exports[`Variable Decorators Converts String correctly 1`] = `
""use strict";

var hello = ___captureAssignment("VariableDeclaration", "hello", 1, "Hello World", "Hello World");"
`;

exports[`Variable Decorators Converts const assignment correctly 1`] = `
""use strict";

var fixedCount = ___captureAssignment("VariableDeclaration", "fixedCount", 1, 8, 8);"
`;

exports[`Variable Decorators Converts let assignment correctly 1`] = `
""use strict";

var count = ___captureAssignment("VariableDeclaration", "count", 1, 1, 1);"
`;

exports[`Variable Decorators Converts multi value assignment correctly 1`] = `
""use strict";

var w = ___captureAssignment("VariableDeclaration", "w", 1, 4, 4),
  x = ___captureAssignment("VariableDeclaration", "x", 1, 3, 3);
var y = ___captureAssignment("VariableDeclaration", "y", 2, undefined, undefined),
  z = ___captureAssignment("VariableDeclaration", "z", 2, 54, 54);"
`;

exports[`Variable Decorators Converts var assignment correctly 1`] = `
""use strict";

var z = ___captureAssignment("VariableDeclaration", "z", 1, 54, 54);"
`;
